src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,146 @@
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index >= size + 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to add element to non exist index"
+            );
+        }
+        if (isFull()) {
+            ensureCapacity();
+        }
+        T[] newElements = (T[]) new Object[elements.length];
 
+        for (int i = 0; i <= size; i++) {
+            if (i > index) {
+                newElements[i] = elements[i - 1];
+            } else {
+                newElements[i] = elements[i];
+            }
+        }
+        newElements[index] = value;
+        size++;
+        elements = newElements;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException("You"
+                    + " tried to get element by non existed index");
+        }
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to set element to non exist index"
+            );
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to remove element from non exist index"
+            );
+        }
+        final T elementToRemove = elements[index];
+        T[] newElements = (T[]) new Object[elements.length];
+        for (int i = 0; i < size; i++) {
+            if (i > index) {
+                newElements[i - 1] = elements[i];
+            } else {
+                newElements[i] = elements[i];
+            }
+        }
+        elements = newElements;
+        size--;
+        return elementToRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T[] newElements = (T[]) new Object[elements.length];
+        T elementToRemove = null;
+        boolean isElementFound = false;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elements[i])) {
+                isElementFound = true;
+                elementToRemove = elements[i];
+            }
+            if (isElementFound) {
+                newElements[i] = elements[i + 1];
+            } else {
+                newElements[i] = elements[i];
+            }
+        }
+        elements = newElements;
+        if (isElementFound) {
+            size--;
+            return elementToRemove;
+        } else {
+            throw new NoSuchElementException();
+        }
```

Comment:
In the `remove(T element)` method, you are creating a new array and copying elements one by one. This is not necessary and inefficient. You can use `System.arraycopy()` to shift the portion of your array to the left of the index where you found the element. This way, you only need to decrease the `size`. Also, the `Objects.equals()` method you used is part of the `java.util.Objects` class which is not allowed according to the checklist.